tree/treeplayer/src/TTreeReader.cxx

```diff
@@ -411,14 +421,35 @@ TTreeReader::EEntryStatus TTreeReader::SetEntryBase(Long64_t entry, Bool_t local
       for (size_t i = 0; i < fValues.size(); ++i) {
          ROOT::Internal::TTreeReaderValueBase* reader = fValues[i];
          reader->CreateProxy();
-
          if (!reader->GetProxy()){
             fEntryStatus = kEntryDictionaryError;
             return fEntryStatus;
          }
+
       }
       // If at least one proxy was there and no error occurred, we assume the proxies to be set.
       fProxiesSet = !fValues.empty();
+
+      // Now we need to properly set the TTreeCache. We do this in steps:
+      // 1. We drop all branches and reset the cache, in case it's not the first time we
+      //    are here.
+      // 2. We set the entry range according to the entry range of the TTreeReader
+      // 3. We add to the cache the branches identifying them by the name the user provided
+      //    upon creation of the TTreeReader{Value, Array}s
+      // 4. We stop the learning phase.
+      // Operations 2, 3 and 4 need to happen in this order. See: https://sft.its.cern.ch/jira/browse/ROOT-9773?focusedCommentId=87837
+      if (fProxiesSet) {
+         if (const auto curFile = fTree->GetCurrentFile()) {
+            // We need to avoid to pass -1 as end entry to the SetCacheEntryRange method
+            const auto lastEntry = (-1LL == fEndEntry) ? fTree->GetEntriesFast() : fEndEntry;
+            fTree->SetCacheEntryRange(fBeginEntry, lastEntry);
```

Comment:
One things to test/check is the current code behavior (i.e. it might need to be fixed or hopefully is already fine) when GetEntriesFast return TTree::kMaxEntries (i.e. in case of TChain).  It should work but I have not tested it.

Reply:
I suspect test with a TChain with one file and then a TChain with 2 files should be sufficient.